### ISSUE TITLE
Using the adjacency matrix as a boolean mask

### ISSUE THREAD

**Issue by JackMaguire** [danielegrattarola/spektral#103]
Hi @danielegrattarola ,

I have a total newbie question. I'm currently running my edge feature matrix through some layers that do not enforce that 0's stay as 0's. Is there a way to use the adjacency matrix as a mask to re-zero out non-edges?

To clarify, suppose my edge feature matrix E looked like this:

```
[ [ 0, 2, 0 ],
  [ 2, 0, 4 ],
  [ 0, 4, 0 ] ]
```

and Adjacency matrix looked like this:

```
[ [ 0, 1, 0 ],
  [ 1, 0, 1 ],
  [ 0, 1, 0 ] ]
```

E may end up looking like this after a few layers:

```
[ [ 8, 3, 2 ],
  [ 3, 8, 1 ],
  [ 2, 1, 8 ] ]
```

Is there a layer or operation that takes E and A and spits out this?

```
[ [ 0, 3, 0 ],
  [ 3, 0, 1 ],
  [ 0, 1, 0 ] ]
```

Any advice is appreciated!
-Jack

**Comment by danielegrattarola** [comment]
Hi Jack, 

a few ideas come to mind. I'm going to assume that E is 2-dimensional, as in your example, although the solutions should work even if E is 3D. 

- If A and E are dense: `A * E`
- If A is sparse, E is dense: `tf.sparse.to_dense(A) * E`
- If A and E are both sparse it's a bit more tricky: 
```python
eq_0 = tf.equal(E.indices[:, 0, None], A.indices[:, 0])
eq_1 = tf.equal(E.indices[:, 1, None], A.indices[:, 1])
eq_and = tf.logical_and(eq_0, eq_1)
mask = tf.reduce_any(eq_and, axis=1)
values = tf.boolean_mask(E_sp.values, mask)
out = tf.SparseTensor(A_sp.indices, values, A_sp.dense_shape)
```
This has a memory cost of O(num_edges_A * num_edges_E), but it only allocates a boolean matrix of that size so it should not have too much impact. 

Cheers, 
Daniele

**Comment by JackMaguire** [comment]
Thanks for the great reply! The tensors are all dense for me at the moment and I can't believe the answer is as simple as `A*E`. This will really help me out.

**Comment by JackMaguire** [comment]
Feel free to ignore this but I still don't quite understand how this works. [Wolfram doesn't give the masking output that I would have expected](https://www.wolframalpha.com/input/?i=%5B+%5B+1%2C+2+%5D%2C+%5B+2%2C+1+%5D+%5D+*+%5B+%5B+0%2C+1+%5D%2C+%5B+1%2C+0%5D+%5D).

**Comment by danielegrattarola** [comment]
I think that is because Wolfram interprets * as matrix multiplication (tf.matmul), while in Python it usually means element-wise product (tf.multiply).
Same thing with numpy and pytorch. 

Cheers, 
Daniele

**Comment by JackMaguire** [comment]
That makes sense. Thanks so much!

**Comment by danielegrattarola** [comment]
Happy to help!!
